### PR TITLE
Throwing exception when resource contains multiple cristin identifiers

### PR DESCRIPTION
**Pull request by truhacevkir**
- Adding custom MultipleCristinIdentifiersException when attempting to persist publication/resource with multiple Cristin identifiers.
- Returning BadRequest when attempting to create publication with multiple Cristin identifier via api. 

For now resourceService throws following when persisting publication with multiple Cristin identifiers, as Cristin identifier serves as unique PK4:

`"java.lang.IllegalArgumentException: Expected zero or a single value, but 2 were found (through reference chain: no.unit.nva.publication.model.storage.ResourceDao[\"PK4\"]) at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4618) at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:4549) at no.unit.nva.publication.model.storage.DataCompressor.lambda$compressDaoData$0(DataCompressor.java:55) at nva.commons.core.attempt.Try.attempt(Try.java:33)`

### CODE REVIEW

File: publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java [hunk 1]
```diff
@@ -142,7 +142,27 @@ private static Resource convertToResource(Publication publication) {
...
+
+    private Resource validate() {
+        var cristinIdentifiers = getCristinIdentifiers();
+        if (hasMultipleEntries(cristinIdentifiers)) {
+            throw new MultipleCristinIdentifiersException("Multiple cristin identifiers are not allowed!");
+        } else {
+            return this;
+        }
+    }
+
```
Comment (truhacevkir) [inline publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java:156]: Adding following validation to `Resource.fromPublication()` method so we always validate dao to persist.

File: publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java [hunk 2]
```diff
@@ -142,7 +142,27 @@ private static Resource convertToResource(Publication publication) {
...
+                   .build().validate();
+    }
+
+    private Resource validate() {
+        var cristinIdentifiers = getCristinIdentifiers();
+        if (hasMultipleEntries(cristinIdentifiers)) {
+            throw new MultipleCristinIdentifiersException();
+        } else {
+            return this;
+        }
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java:154]: I'm not sure that this is really the case — are there zero duplicates in Cristin?

File: publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java [hunk 3]
```diff
@@ -142,7 +142,27 @@ private static Resource convertToResource(Publication publication) {
...
+    }
+
+    private static boolean hasMultipleEntries(List<String> cristinIdentifiers) {
+        return cristinIdentifiers.size() > 1;
+    }
+
+    private List<String> getCristinIdentifiers() {
+        return this.getAdditionalIdentifiers().stream()
+                                    .map(AdditionalIdentifier::getSourceName)
+                                    .filter("Cristin"::equals)
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java:164]: Constant?

File: publication-commons/src/test/java/no/unit/nva/publication/service/impl/ResourceServiceTest.java [hunk 4]
```diff
@@ -1187,6 +1188,16 @@ void shouldThrowBadRequestWhenAttemptingToDeletePublishedPublication() throws Ap
                      () -> resourceService.deletePublication(resourceService.getPublication(publication)));
     }
 
+    @Test
+    void shouldThrowExceptionWhenPersistingPublicationWithMultipleCristinIds() {
+        var publication = randomPublication();
+        publication.setAdditionalIdentifiers(Set.of(new AdditionalIdentifier("Cristin", randomString()),
+                                                    new AdditionalIdentifier("Cristin", randomString())));
```
Comment (brinxmat) [inline publication-commons/src/test/java/no/unit/nva/publication/service/impl/ResourceServiceTest.java:1195]: Constant

File: publication-rest/src/main/java/no/unit/nva/publication/create/CreatePublicationHandler.java [hunk 5]
```diff
@@ -110,13 +117,25 @@ protected PublicationResponse processInput(CreatePublicationRequest input, Reque
...
 
-        return PublicationResponse.fromPublication(createdPublication);
+        return publicationResponse;
     }
 
+    private ApiGatewayException mapException(Failure<PublicationResponse> failure) {
+        return switch (failure.getException()) {
+            case NotFoundException ex -> ex;
+            case UnauthorizedException ex -> ex;
+            case MultipleCristinIdentifiersException ex -> new BadRequestException(MULTIPLE_CRISTIN_IDENTIFIERS_MESSAGE);
```
Comment (brinxmat) [inline publication-rest/src/main/java/no/unit/nva/publication/create/CreatePublicationHandler.java:133]: Missing BadRequestException, possibly others we aren't testing. This causes `CreatePublicationHandlerTest.shouldReturnBadRequestWhenAnExternalClientTriesToCreatePublishedDocumentWithoutTitleAndDoiRef` to fail

File: publication-rest/src/test/java/no/unit/nva/publication/create/CreatePublicationHandlerTest.java [hunk 6]
```diff
@@ -553,6 +555,20 @@ void shouldValidateOkIfInstanceTypeIsNotSet() throws IOException {
         assertThat(response.getStatusCode(), is(equalTo(HttpURLConnection.HTTP_CREATED)));
     }
 
+    @Test
+    void shouldReturnBadRequestWhenAttemptingToPersistPublicationWithMultipleCristinIdentifiers() throws Exception {
+        var request = createEmptyPublicationRequest();
+        request.setAdditionalIdentifiers(Set.of(new AdditionalIdentifier("Cristin", randomString()),
+                                                new AdditionalIdentifier("Cristin", randomString())));
```
Comment (brinxmat) [inline publication-rest/src/test/java/no/unit/nva/publication/create/CreatePublicationHandlerTest.java:562]: I think the real issue here is that AdditionalIdentifiers, or at least a subset of these are in fact their own entries in DDB. This allows us to look them up and relate them to the publication, e.g.

```
  | publicationid | publicationid | publicationStuff |
  | publicationid | handle        |  |
  | publicationid | handle        |  |
  | publicationid | cristinId     |  |
  | publicationid | doi           |  |
```

File: publication-rest/src/test/java/no/unit/nva/publication/create/CreatePublicationHandlerTest.java [hunk 7]
```diff
@@ -553,6 +555,20 @@ void shouldValidateOkIfInstanceTypeIsNotSet() throws IOException {
...
+        var inputStream = createPublicationRequest(request);
+        handler.handleRequest(inputStream, outputStream, context);
+
+        var actual = GatewayResponse.fromOutputStream(outputStream, Problem.class);
+        assertThat(actual.getStatusCode(), is(equalTo(HttpURLConnection.HTTP_BAD_REQUEST)));
+        var publicationResponse = actual.getBodyObject(Problem.class);
+        assertThat(publicationResponse.getDetail(), containsString("Multiple Cristin identifiers are not allowed"));
```
Comment (brinxmat) [inline publication-rest/src/test/java/no/unit/nva/publication/create/CreatePublicationHandlerTest.java:569]: The fact we have more than one assertion here makes me wish we used a testing framework that supports rest testing — not your issue, but we should look into this.

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java (hunk 2 above)
Comment (truhacevkir) [reply publication-commons/src/main/java/no/unit/nva/publication/model/business/Resource.java:154]: When importing publications from Brage, multiple publications are ending up with IllegalArgumentException when we are trying to persist publication with multiple Cristin id's. There should not be duplicates in Cristin or Brage, but I think we have to handle cases when trying to persist publication with multiple Cristin identifiers.

**Review by brinxmat — APPROVED** [review]

**Pull request merged**